src/usr.sbin/flashcfg/obs266/flashcfg.8
.\"	$ssdlinux: flashcfg.8,v 1.1 2004/11/09 07:54:28 yamagata Exp $
.\"
.\" Copyright (c) 2003-2004
.\"         Plat'Home CO., LTD. <openlab@plathome.co.jp>. All rights reserved.
.\"
.\" Conditions and the no-warranty notice below apply to this software except
.\" software which has its own conditions and the no-warranty notice.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.TH FLASHCFG "8" "April 2003" "OpenBlockS266" "System management commands"
.SH NAME
flashcfg \- manage FlashROM
.SH SYNOPSIS
.B flashcfg
\fB\-c \fIrootdev\fR
.fi
.PP
.B flashcfg
\fB\-d\fR
.fi
.PP
.B flashcfg
\fB\-f \fIboot_image\fR
.fi
.PP
.B flashcfg
\fB\-h\fR
.fi
.PP
.B flashcfg
\fB\-s \fIlist_file\fR
.fi
.PP
.B flashcfg
\fB\-x\fR
.fi
.SH DESCRIPTION
The flashcfg command restores files and directories from FlashROM,
saves them to it, deletes them from it,
loads a boot image to it and changes the root file system.
.TP
\fB\-c \fIrootdev\fR
change the root file system to \fIrootdev\fR: \fIinitrd\fR (initrd),
\fIharddisk\fR (hard disk), \fIhda1,...,had4\fR (partition)
.TP
\fB\-d\fR
delete files saved from FlashROM
.TP
\fB\-f \fIboot_image\fR
load the boot image, \fIboot_image\fR, to FlashROM
.TP
\fB\-h\fR
display this help and exit
.TP
\fB\-s \fIlist_file\fR
save files and directoties in \fIlist_file\fR (/etc/flashcfg) to FlashROM
.TP
\fB\-x\fR
restore files and directoties from FlashROM
.SH FILES
.TP
.I /etc/flashcfg
list of files and directories restored or saved.
.SH AUTHOR
Written by Plat'Home CO., LTD. <openlab@plathome.co.jp>.
.SH "REPORTIONG BUGS"
Report bugs to <openlab@plathome.co.jp>.
.SH COPURIGHT
Copyright \(co 2003-2004 Plat'Home CO., LTD.


src/ChangeLog
$ssdlinux: ChangeLog,v 1.492 2005/03/20 09:18:33 kanoh Exp $

Sun Mar 20 18:15:57 JST 2005  Tamotsu Kanoh <kanoh@kanoh.org>
	* Update GCC to version 3.3.5.

Fri Mar 18 11:27:00 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update findutils to version 4.2.20.

Fri Mar 18 09:57:30 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050318 Release.

Wed Mar 16 17:00:49 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update strace to version 4.5.10.

Wed Mar 16 15:48:34 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update postfix to version 2.2.1.

Tue Mar 15 11:32:34 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update gettext to version 0.14.3.

Tue Mar 15 10:18:52 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update psmisc to version 21.6.

Mon Mar 14 12:48:35 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Updated bind to version 9.2.5.

Mon Mar 14 10:15:00 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050312.

Fri Mar 11 16:43:48 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update tnftp to version 20050103.

Fri Mar 11 09:58:54 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050311 Release.

Thu Mar 10 17:44:49 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update postfix to version 2.2.0.

Thu Mar 10 16:11:46 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update iptables to version 1.3.1.

Thu Mar 10 14:14:08 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update openssh to version 4.0p1.

Mon Mar  7 10:15:38 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050305.

Fri Mar  4 18:12:35 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050304 Release.

Wed Mar  2 15:04:15 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update e2fsprogs to version 1.36.

Wed Mar  2 10:45:01 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update dhcp to version 3.0.2.

Mon Feb 28 10:14:46 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050226.

Fri Feb 25 16:23:38 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ex to version 050225.

Fri Feb 25 16:05:29 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update gettext to version 0.14.2.

Fri Feb 25 09:46:36 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050225 Release.

Mon Feb 21 11:58:44 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update zlib to version 1.2.2.

Mon Feb 21 11:28:09 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update findutils to version 4.2.18.

Mon Feb 21 10:57:22 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050219.

Fri Feb 18 10:39:24 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050218 Release.

Tue Feb 15 18:33:14 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050215 Release.

Mon Feb 14 12:03:37 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update iptables to version 1.3.0.

Mon Feb 14 11:03:45 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update file to version 4.13.

Mon Feb 14 10:40:03 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update libtool to version 1.5.14.

Mon Feb 14 10:33:00 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update automake to version 1.9.5.

Mon Feb 14 10:28:42 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050212.

Thu Feb 10 18:02:01 JST 2005  Jun Todoroki <todoroki@plathome.co.jp>
	* Update usbustils to version 0.70 & add libusb-0.1.9.

Wed Feb  9 11:34:42 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update grub to version 0.96.

Wed Feb  9 14:36:01 JST 2005  Jun Todoroki <todoroki@plathome.co.jp>
	* Update madwifi to version 20050209.

Mon Feb  7 14:20:34 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update strace to version 4.5.9.

Mon Feb  7 13:47:22 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update sudo to version 1.6.8p7.

Mon Feb  7 11:17:34 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update cvs to version 1.11.19.

Mon Feb  7 10:47:05 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update libtool to version 1.5.12.

Mon Feb  7 10:38:15 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050205.

Mon Jan 31 12:29:41 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update sed to version 4.1.4.

Mon Jan 31 12:19:56 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050129.

Mon Jan 31 11:29:54 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update findutils to version 4.2.15.

Sat Jan 29 18:05:07 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050129 Release.

Thu Jan 27 15:57:02 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update util-linux to version 2.12q.

Thu Jan 27 12:41:46 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update shadow to version 4.0.7.

Thu Jan 27 11:38:48 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update procps to version 3.2.5.
	
Thu Jan 27 11:18:56 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ex to version 050119.

Thu Jan 27 11:01:33 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update dump to version 0.4b39.

Thu Jan 27 10:36:16 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update sed to version 4.1.3.

Wed Jan 26 17:08:50 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050122.

Fri Jan 21 23:18:22 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050121 Release.

Wed Jan 19 11:18:33 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update a patch for sendmail (Errata: 2005-01-17).

Mon Jan 17 11:04:35 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
        * Apply a patch for sendmail (Errata: 2005-01-15).


src/usr.sbin/flashcfg/obs266/flashcfg.c
/*	$ssdlinux: flashcfg.c,v 1.18 2004/11/02 02:19:36 todoroki Exp $	*/
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <fcntl.h>
#include <time.h>
#include <sys/ioctl.h>
#include <sys/mount.h>
#include <sys/stat.h>
#include <linux/mtd/mtd.h>

extern char **environ;
/* some variables used in getopt (3) */
extern char *optarg;
extern int optind;
extern int optopt;
extern int opterr;
extern int optreset;

#define	KILO_BYTE	1024
#define	MEGA_BYTE	(1024*1024)
#define	SECT1_SIZE	(8*KILO_BYTE)
#define	SECT2_SIZE	(64*KILO_BYTE)
#define	FLASH_SIZE	(8*MEGA_BYTE)
#define	MAGIC		0x0052504F

#define STATUS_COL	50

typedef struct nvram_boot_t {
	char boot_magic[2];
	char boot_dev[14];
} nvram_boot_t;

nvram_boot_t nvram_boot;

#define	NVRAM_BOOT_OFF	256

typedef struct boot_block_t {
	unsigned long	magic;
	unsigned long	dest;
	unsigned long	num_512blocks;
	unsigned long	debug_flag;
	unsigned long	entry_point;
	unsigned long	reserved[3];
} boot_block_t;

boot_block_t * pboot_block;

unsigned short *membase;

void flash_set_bootdev(char *);
void flash_prog_kern(char *);
void flash_save_param(char *);
void flash_delete_param(void);
void flash_extract_param(void);

/*#define DEBUG*/
#define ENABLE_MULTIBOOT

void
usage()
{
	fprintf(stderr, "usage: flashcfg                Show this.\n");
#ifdef	ENABLE_MULTIBOOT
	fprintf(stderr, "       flashcfg -c [rootdev]   Change root file system [initrd|harddisk|hda1-4]\n");
#else
	fprintf(stderr, "       flashcfg -c [rootdev]   Change root file system [initrd|harddisk]\n");
#endif
	fprintf(stderr, "       flashcfg -f boot_image  Load boot image to flash\n");
	fprintf(stderr, "       flashcfg -s list_file   Save files to flash\n");
	fprintf(stderr, "       flashcfg -x             Restore files from flash\n");
	fprintf(stderr, "       flashcfg -d             Delete saved files from flash\n");
	fprintf(stderr, "       flashcfg -h             Show this.\n");
}

int
main(int argc, char *argv[])
{
	FILE *fp;
	int i, f;
	char *boot_dev;

	if (getuid()) {
		fprintf(stderr, "must be super user\n");
		return (0);
	}

	while ((i = getopt(argc, argv, "c:f:s:xdh")) != -1) {
		switch (i) {
		case 'c':
			if (strcmp(optarg, "harddisk") == 0)
				boot_dev = "\001\007hda1";
			else if (strcmp(optarg, "initrd") == 0)
				boot_dev = "\002\007ram";
#ifdef  ENABLE_MULTIBOOT
			else if (strcmp(optarg, "hda1") == 0)
				boot_dev = "\001\007hda1";
			else if (strcmp(optarg, "hda2") == 0)
				boot_dev = "\001\007hda2";
			else if (strcmp(optarg, "hda3") == 0)
				boot_dev = "\001\007hda3";
			else if (strcmp(optarg, "hda4") == 0)
				boot_dev = "\001\007hda4";
#endif
			else {
				fprintf(stderr, "invalid option %s\n",
					optarg);
				return (1);
			}
#ifdef DEBUG
			fprintf(stderr, "option %c: arg = %s\n", i, optarg);
#endif
			flash_set_bootdev(boot_dev);
                        fprintf(stderr, "Boot device change to %s\n", optarg);
			return (0);

		case 'f':
#ifdef DEBUG
			fprintf(stderr, "option %c: arg = %s\n", i, optarg);
#endif
			flash_prog_kern(optarg);
                        fprintf(stderr, "done\n");
			return (0);

		case 's':
#ifdef DEBUG
			fprintf(stderr, "option %c: arg = %s\n", i, optarg);
#endif
			flash_save_param(optarg);
                        fprintf(stderr, "done\n");
			return (0);

		case 'x':
			flash_extract_param();
			return (0);

		case 'd':
			flash_delete_param();
                        fprintf(stderr, "done\n");
			return (0);

		case 'h':
			usage();
			return (0);

		default:
			usage();
			return (0);
		}
	}
        if (argc != 1) {
                usage();
                return (0);
        }
	argc -= optind;
	argv += optind;
/*	flash_extract_param(); */
	usage();
	return (0);
}

void
flash_set_bootdev(char *arg)
{
	int fd, n, i;

#ifdef DEBUG
	printf("boot device option = %s\n", arg);
#endif
	if ((fd = open("/dev/nvram0", O_RDWR)) < 0) {
		perror("/dev/nvram0");
		return;
	}

	if (lseek(fd, NVRAM_BOOT_OFF, 0L) == -1) {
		perror("lseek");
		return;
	}

	if ((n = read(fd, &nvram_boot, sizeof (nvram_boot))) !=
		sizeof (nvram_boot)) {
		perror("read");
		close(fd);
		return;
	}

	nvram_boot.boot_magic[0] = 'N';
	nvram_boot.boot_magic[1] = 'V';
	(void)strcpy(nvram_boot.boot_dev, arg);

	if (lseek(fd, NVRAM_BOOT_OFF, 0L) == -1) {
		perror("lseek");
		return;
	} 

	if ((i = write(fd, &nvram_boot, n)) != n) {
		perror("write");
	}
	close(fd);
}

void
flash_prog_kern(char *kern)
{
	int i, st, ifd, ofd, nread;
	struct stat sb;
	erase_info_t erase;

	if ((ifd = open(kern, O_RDONLY)) < 0) {
		perror(kern);
		return;
	}

	if (fstat(ifd, &sb) < 0) {
		perror("stat");
		close(ifd);
		return;
	}

	if (sb.st_size > (FLASH_SIZE - 8*SECT1_SIZE - 256*KILO_BYTE)) {
		fprintf(stderr, "file size exceeds MAX kernel SIZE\n");
		close(ifd);
		return;
	}

	if ((membase = malloc(SECT2_SIZE)) == NULL) {
		perror("malloc");
		return;
	}

	nread = read(ifd, membase, SECT2_SIZE);

	if (nread < SECT2_SIZE) {
		fprintf(stderr, "file isn't kernel image\n");
		free(membase);
		close(ifd);
		return;
	}

	pboot_block = (boot_block_t *)membase;
	
	if (pboot_block->magic != MAGIC) {
		fprintf(stderr, "file isn't kernel image\n");
		free(membase);
		close(ifd);
		return;
	}

	if ((ofd = open("/dev/mtd/1", O_RDWR)) < 0) {
		perror("/dev/mtd/1");
		close(ifd);
		return;
	}

	erase.start = 0;
	erase.length = SECT2_SIZE;
	i=0;
	lseek(ifd, 0, SEEK_SET);
	fprintf(stderr, "Load boot image to FlashROM\n");
	while ((nread = read(ifd, membase, SECT2_SIZE)) > 0) {
		fprintf(stderr, ".");
		if (ioctl(ofd, MEMERASE, &erase) != 0) {
			perror("Erase failure");
			close(ifd);
			close(ofd);
			return;
		}
		fprintf(stderr, "%c#",0x08);
		if (write(ofd, membase, nread) < 0) {
			perror("mtd");
			close(ifd);
			close(ofd);
			return;
		}
		i++;
		erase.start += nread;
		if(i>=STATUS_COL) {
			fprintf(stderr, "\n");
			i=0;
		}
	}
	fprintf(stderr, "\n");

	free(membase);
	close(ifd);
	close(ofd);
}

void
flash_save_param(char *list)
{
	int i, pid, st, ifd, ofd, nread;
	struct stat sb;
	char localbuf[256];
	erase_info_t erase;

	strcpy(localbuf, "/tmp/flashcfg.XXXXXX");

	if ((ifd = mkstemp(localbuf)) < 0) {
		perror("mkstemp");
		return;
	}

	if ((pid = fork()) == 0) {
		/* in child */
		execl("/usr/bin/tar", "tar", "cvpTzf", list, localbuf, NULL);
		/* not reached here */
	} else {
		wait(&st);
	}

#ifdef DEBUG
	if (st)
		fprintf(stderr, "child returns %d\n", st);
#endif

	if ((ifd = open(localbuf, O_RDONLY)) < 0) {
		perror(localbuf);
		return;
	}

	if (fstat(ifd, &sb) < 0) {
		perror("stat");
		close(ifd);
		return;
	}

	if (sb.st_size > 8 * SECT1_SIZE) {
		fprintf(stderr, "file size exceeds 64KB\n");
		close(ifd);
		return;
	}

	if ((membase = malloc(SECT1_SIZE)) == NULL) {
		perror("malloc");
		return;
	}

	if ((ofd = open("/dev/mtd/0", O_RDWR)) < 0) {
		perror("/dev/mtd/0");
		close(ifd);
		return;
	}

	erase.start = 0;
	erase.length = SECT1_SIZE;
	i=0;
	fprintf(stderr, "Save files to FlashROM\n");
	while ((nread = read(ifd, membase, SECT1_SIZE)) > 0) {
		fprintf(stderr, ".");
		if (ioctl(ofd, MEMERASE, &erase) != 0) {
			perror("Erase failure");
			close(ifd);
			close(ofd);
			return;
		}
		fprintf(stderr, "%c#",0x08);
		if (write(ofd, membase, nread) < 0) {
			perror("mtd");
			close(ifd);
			close(ofd);
			return;
		}
		erase.start += nread;
		if(i>STATUS_COL) {
			fprintf(stderr, "\n");
			i=0;
		}
		i++;
	}
	fprintf(stderr, "\n");

	free(membase);
	close(ifd);
	close(ofd);
	unlink(localbuf);
}

void
flash_extract_param()
{
	int pid, st;

	if ((pid = fork()) == 0) {
		/* in child */
		execl("/usr/bin/tar", "tar", "--no-time-check",
			"-xvpzf", "/dev/mtd/0",
			"-C", "/", NULL);
		/* not reached here */
	} else {
		wait(&st);
	}
#ifdef DEBUG
	if (st)
		fprintf(stderr, "child returns %d\n", st);
#endif
}

void flash_delete_param()
{
	int ofd, i;
	erase_info_t erase;

	if ((ofd = open("/dev/mtd/0", O_RDWR)) < 0) {
		perror("/dev/mtd/0");
		close(ofd);
		return;
	}

	erase.start = 0;
	erase.length = SECT1_SIZE;
	fprintf(stderr, "Delete saved files from FlashROM\n");
	for (i = 0; i < 8; i++) {
		fprintf(stderr, ".");
		fflush(stderr);
		if (ioctl(ofd, MEMERASE, &erase) != 0) {
			perror("Erase failure");
			close(ofd);
			return;
		}
		erase.start += SECT1_SIZE;
	}
	fprintf(stderr, "\n");
	close(ofd);
}
